### DISCUSSION TITLE
[signalStore] How to refer to the type of a particular store?

### DISCUSSION THREAD

**Discussion by jits** [ngrx/platform#4140] [Q&A] [answered]
How can I refer to the _type_ of a particular store?

E.g. if I want to pass the store as an argument into a function:

```ts
type AuthState = { user: User | null };

export const AuthStore = signalStore(
  { providedIn: 'root' },
  withState<AuthState>({ user: null })
);

const aHelperFunction = (store: ???) => {
   // Here I want to be able to access `store.user()` (and anything else published by the store, like computed signals and methods).
}
```

I've tried `store: typeof AuthStore` but this gives a type of `Type<…>`, which seems to be something from `@angular/core` — _"Represents a type that a Component or other object is instances of."_ — and there doesn't seem to be a way to refer to the "internals" of the store from this type.

I suspect I'm missing some knowledge here on a) how the signal store is modelled (with an inner `SignalStore` class) and/or b) something to do with signals and injection context meaning we can't easily use the store like this, in arbitrary functions.

**Comment by markostanimirovic** [comment, marked as answer]
You can define the type of the SignalStore in the following way:

```ts
// users.store.ts

export const UsersStore = signalStore(
  withState({ users: [] as User[] })
);

export type UsersStore = InstanceType<typeof UsersStore>;
```

**Note:** Exported type has the same name as the `UsersStore` constant. This is not required, but when defined like this, imported `UsersStore` can be used as a type as well as value:

```ts
// 1) as a type:
import { UsersStore } from './users.store';

function doSomething(usersStore: UsersStore) {}

// 2) as a value:
import { UsersStore } from './users.store';

class UsersComponent {
  readonly usersStore = inject(UsersStore);
}

// 3) type + value:
import { UsersStore } from './users.store';

@Component({
  providers: [UsersStore], // value
})
class UsersComponent {
  constructor(readonly usersStore: UsersStore) {} // type
}
```

**Reply by jits** [reply]
Great! Thanks for that comprehensive answer, Marko.

**Reply by jits** [reply]
~~When defining the type with the same name as the constant, in the same file, TypeScript complains with the error:~~

```
Duplicate identifier 'AuthStore'.ts(2300)
```

See the comments below.

**Reply by markostanimirovic** [reply]
Here is the example: https://stackblitz.com/edit/stackblitz-starters-3twl8g?file=src%2Fmain.ts

**Reply by jits** [reply]
Thanks @markostanimirovic.

I'm only getting the error when defining the store as a class.

```ts
const _AuthStore = signalStore(…);

@Injectable({ providedIn: 'root' })
export class AuthStore extends _AuthStore {
  …
}

export type AuthStore = InstanceType<typeof AuthStore>;
```

~~So it's a bit of an edge case.~~ See next comment.

**Reply by jits** [reply]
Okay, so it's just me getting myself in a hole with TypeScript 😓 — there's no need for the additional `export type` if you already have a class definition!

Apologies for the unnecessary back and forth.

**Comment by Pizzicato** [comment]
Hi! This works well when the store instance has been created, but how could it be done before that, for example, if I need to access the store type from a function that returns an `rxMethod` depending on a parameter, and that function is used inside one of the `withMethods` declaration?

**Reply by rainerhahnekamp** [reply]
@Pizzicato do you have a small example for that? pseudocode should be enough. Do you mind open a new discussion for that one?

**Reply by Pizzicato** [reply]
Sorry for the late answer, I didn't get a notification.

Yep, sure, here's the code: 

```typescript
export const MyStore = signalStore(
  ...
  withMethods((store, myService = inject(MY_SERVICE, { optional: true })) => ({
    search: myService ? asyncMethod(myService, store) : syncMethod(store)
  })),
  ...
```

`asyncMethod` and `syncMethod` both need the `store` instance, and I don't know how to type it in the functions definition.

Thanks a lot!

**Reply by rainerhahnekamp** [reply]
You don't get the type. What you have to do instead, is that your asyncMethod defines the methods or signals it needs to call.

So if your store exposes a load method your `asyncMethod` wants to use, it is going to be:

```typescript
function asyncMethod(myService: MY_SERVICE, partialStore: {load: (id: number) => Promise<void>) {

}
```

There is a similar issue going on in https://github.com/ngrx/platform/discussions/4536 (@devkirr, you might want to mark your question as answered, if it fulfills your needs)

**Reply by Pizzicato** [reply]
Hmm, the problem is that in those methods I need to run `patchState`. So I need the full store.

**Reply by rainerhahnekamp** [reply]
Okay, we should be able to work something out, but I'd like to ask two favours of you:

1. Can you please come up with stackblitz example with a minimal store and your `asyncMethod`?
2. Can you please open a new GitHub issue. In case we need for reference, it is easier if we have a dedicated one for your question.

**Reply by Pizzicato** [reply]
Hi, sure, I'll do that this week. I'm really busy and need some spare time to find. Thanks for your help.

**Reply by rainerhahnekamp** [reply]
Here's the stackblitz template: https://stackblitz.com/github/ngrx/signal-store-starter

**Reply by devkirr** [reply]
@Pizzicato add the following type to the store definition passed into your function

```ts
type Store = {
  load: (id: number) => Promise<void>
} & StateSignal<TypeOfYourState>

function asyncMethod(myService: MY_SERVICE, partialStore: Store) {

}
```

I did that to be able to pass a partial Store into a directive `@Input()` to access state and call store methods and it worked well.

**Reply by Pizzicato** [reply]
@devkirr thanks for the suggestion! I tried and it seems to return typescript errors when using a store with that type in the `patchState` function, so I guess that's an approximation to the Store type, but not a full one.

**Reply by Pizzicato** [reply]
@rainerhahnekamp I found some time to fulfill your requests:
1. [Stackblitz](https://stackblitz.com/edit/github-crxqjl?file=src%2Fstore.ts)
2. [Issue](https://github.com/ngrx/platform/issues/4547)

Thanks a lot! BTW I really enjoy watching your YouTube channel, really nice and last minute content, keep it up! 💟

**Reply by devkirr** [reply]
> @devkirr thanks for the suggestion! I tried and it seems to return typescript errors when using a store with that type in the `patchState` function, so I guess that's an approximation to the Store type, but not a full one.

@Pizzicato you're right, StateSignal worked for `getState` method but to use `patchState` you need the type `WritableStateSource` instead

**Reply by markostanimirovic** [reply]
https://github.com/ngrx/platform/issues/4547#issuecomment-2397749647
